### PR TITLE
Text to centreline docs + folder restructure

### PR DESCRIPTION
**Pull request by chelsrosic**
starting documentation of process

### CODE REVIEW

File: gis/text_to_centreline_geometry_scripts/text_to_centreline_script_xml_opendata.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Thu Feb 28 10:16:47 2019
+
+Script to convert text from an xml to geometries 
+and put the output into a postgres DB table 
+
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/text_to_centreline_script_xml_opendata.py]: this filename should be simplified to `bylaw_opendata_xml.py`

File: gis/text_to_centreline_geometry_scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
+
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/README.md]: this folder should be renamed to bylaw_text_to_centreline

File: gis/text_to_centreline_geometry_scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+The function outputs a confidence level (`con`), which is defined after the `centreline_segments` variable. This value represents how close the names of the input values are to the intersections they were matched to. Specifically, this value is the sum of the levenshtien distance between the `highway2` value and the street name it matched to and the `btwn1`/`btwn2` value and the street name it matched to. It is the sum of this value for both intersections.
+
+# Special Cases
+
+There are some records that start and/or end at locations that are a certain amount of metres away from an intersection. We create a method to assign a geometry to the locations of these bylaws. These segments do not start at one intersection and end at another, they often start/end in the middle of centreline segments.
+
+### Explore the data/cases
+
+**Case 1**
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/README.md]: The cases can be summarized as:
- "Two intersections"
- "One Intersection + offset"
- "Two intersections and at least one offset"
- "Offset streets" (the multi-linestring issue).

File: gis/text_to_centreline_geometry_scripts/centreline_automation.sql [hunk 4]
```diff
@@ -0,0 +1,1018 @@
+CREATE OR REPLACE FUNCTION crosic.get_intersection_id(highway2 TEXT, btwn TEXT, not_int_id INT)
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/centreline_automation.sql]: This file should be split into individual files for each function.

File: gis/text_to_centreline_geometry_scripts/README.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
+
+# <center>Text Description to Centreline Geometry Automation</center>
+
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/README.md]: Insert Table of Contents (can be generated automagically using the VS Code Markdown All in One extension

File: gis/text_to_centreline_geometry_scripts/centreline_automation.sql [hunk 6]
```diff
@@ -0,0 +1,1018 @@
...
+								-- 10* used to be 3
+								ST_DWithin( ST_Transform(s.geom, 26917), ST_BUFFER(line_geom, 3*ST_LENGTH(line_geom), 'endcap=flat join=round') , 10)
+								-- over 90% of centreline segment must be in the buffer
+								AND ST_Length(st_intersection(ST_BUFFER(line_geom, 3*(ST_LENGTH(line_geom)), 'endcap=flat join=round') , ST_Transform(s.geom, 26917))) /ST_Length(ST_Transform(s.geom, 26917)) > 0.9
+							)
+							OR
+							(
+								-- 10 TIMES LENGTH WORKS .... LOOK INTO POTENTIALLY CHANGING LATER
+								COALESCE(metres_btwn1, metres_btwn2) IS NOT NULL AND
+								ST_DWithin(ST_Transform(s.geom, 26917), ST_BUFFER(line_geom, 10*COALESCE(metres_btwn1, metres_btwn2), 'endcap=flat join=round'), 30)
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/centreline_automation.sql]: Use [`GREATEST`](https://devdocs.io/postgresql~9.6/functions-conditional#FUNCTIONS-GREATEST-LEAST) instead of COALESCE here

File: gis/text_to_centreline_geometry_scripts/centreline_automation.sql [hunk 7]
```diff
@@ -0,0 +1,1018 @@
...
+		COALESCE(metres_btwn1, metres_btwn2) IS NULL AND
+		-- 3* could be changed to 10* if you would like
+		ST_DWithin( ST_Transform(s.geom, 26917), ST_BUFFER(line_geom, 3*ST_LENGTH(line_geom), 'endcap=flat join=round') , 10)
+		AND ST_Length(st_intersection(ST_BUFFER(line_geom, 3*(ST_LENGTH(line_geom)), 'endcap=flat join=round') , ST_Transform(s.geom, 26917))) /ST_Length(ST_Transform(s.geom, 26917)) > 0.9
+	)
+	OR
+	(
+		-- 10 TIMES LENGTH WORKS .... LOOK INTO POTENTIALLY CHANGING LATER
+		COALESCE(metres_btwn1, metres_btwn2) IS NOT NULL AND
+		ST_DWithin(ST_Transform(s.geom, 26917), ST_BUFFER(line_geom, 10*COALESCE(metres_btwn1, metres_btwn2), 'endcap=flat join=round'), 30)
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/centreline_automation.sql]: Use [`GREATEST`](https://devdocs.io/postgresql~9.6/functions-conditional#FUNCTIONS-GREATEST-LEAST) instead of COALESCE here

File: gis/text_to_centreline_geometry_scripts/README.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+ - If the fraction returned by ST_LineLocatePoint of the dissolved centreline that the original intersection occurred on was smaller than the fraction that returned by ST_LineLocatePoint represented the closest point on the dissolved centreline to the end of the line from step 1, then the situation would look like this (assuming that the line goes from left to right so points on the left have a lower fraction value):
+![](jpg/figure1_case2.jpg)
+ - In the situation in the above image, to get the fraction of the point x metres away from the original intersection, you would add the `x metres/total length of dissolved centreline` to the fraction returned by `ST_LineLocatePoint` of the original intersection.
+ ![](jpg/figure2_case2.jpg)
+ - In the situation in the above image (assuming the line is drwan from left to right so that the fractions from St_LineLocatePoint of the points on the left are smaller), to get the fraction of the point x metres away from the original intersection, you would subtract the `x metres/total length of dissolved centreline` to the fraction returned by LineLocatePoint of the original intersection.
+
+# MultiLineString Geometries that are Case 2
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/README.md]: Are offset streets only an issue with bylaws where at least one point is offset from an intersection?

Comment (radumas) [reply gis/text_to_centreline_geometry_scripts/README.md]: Also this should be an `##` level header or convert it to an issue?

File: gis/text_to_centreline_geometry_scripts/README.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+3. make a buffer around the line and catch centreline segments in the buffer with the correct street name
+4. dissolve the centreline segments for each bylaw into one geom
+5. cut the dissolved segments to be from the intersection described in `btwn1` column to x metres away from that intersection in the direction specified in `direction_btwn2`.
+
+### Workflow for Special Case 2:
+1. make a point that is x metres away from intersection in the cardinal direction indicated
+2. make a line from the intersection to that point
+3. make a buffer around the line and catch centreline segments in the buffer with the correct street name
+4. dissolve the centreline segments for each community safety zone into one geom
+5.
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/README.md]: The hierarchy of information here is hard to follow, switches from numbered list to bullets. The image filenames read "case1"

File: gis/text_to_centreline_geometry_scripts/README.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+The function is mainly created to process the City of Toronto's [transportation bylaw data](https://open.toronto.ca/dataset/traffic-and-parking-by-law-schedules/). We have already used previous versions of this process for [posted speed limits](https://github.com/CityofToronto/bdit_data-sources/tree/master/gis/posted_speed_limit_update) on streets in the City, locations of [community safety zones](https://github.com/CityofToronto/bdit_vz_programs/tree/master/safety_zones/commuity_safety_zone_bylaws), [turn restrictions](https://github.com/CityofToronto/bdit_vz_programs/blob/master/notebooks/Turn%20Restrictions.ipynb), etc. The function can handle most bylaws, even more advanced cases. Any limitations that we are currently aware of will be listed in the [QC folder README](QC/README.md) of the document. It also should be noted that some bylaws are incorrect (for many reasons, such as a described intersection not existing), and our function cannot coreectly match a lot of these bylaws, since the data is incorrect. Sometimes the function will match the bylaws incorrectly and other times it will return and error.
+
```
Comment (radumas) [inline gis/text_to_centreline_geometry_scripts/README.md]: Include how to use the python script.

File: gis/bylaw_text_to_centreline/README.md [hunk 11]
```diff
@@ -0,0 +1,300 @@
+
```
Comment (radumas) [inline gis/bylaw_text_to_centreline/README.md:1]: This folder should be `text_to_centreline`
The posted speed limit folder should be under this folder structure

File: gis/bylaw_text_to_centreline/README.md [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+# <center>Text Description to Centreline Geometry Automation</center>
```
Comment (radumas) [inline gis/bylaw_text_to_centreline/README.md:2]: replace all the `jchew` schema with `gis` which is where the code will live

File: gis/bylaw_text_to_centreline/README.md [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+There is a [Github Issue](https://github.com/CityofToronto/bdit_data-sources/issues/188) on weird bylaws that haven't been matched from Chelsea's work. \
+For example: \
+i) The streets stated in bylaws are slightly off i.e. they do not intersect or are rather far apart. \
+ii) Directionality matters. Northbound and Southbound route between two intersection are of different centrelines such as the Allen Rd. \
+iii) The error message `line_locate_point: 1st arg isn't a line` returned for 49 distinct id's. This mostly occurs when trying to slice a multilinestring. 
+
+Majority of them are also some issues that I have encountered with my new processes but have not been solved yet. I will list down the issue below and some idea on how we can tackle them.
+
+## pgRouting returns the shortest path but street name different from `highway`
```
Comment (radumas) [inline gis/bylaw_text_to_centreline/README.md:264]: This was solved right?

Comment (jovenc) [reply gis/bylaw_text_to_centreline/README.md:264]: Kinda? Based on this [comment](https://github.com/CityofToronto/bdit_data-sources/issues/268#issuecomment-620352090), previously there were 84 bylaws which fell under this category, and I solved 69 of them but there are still 15 of them that still fails due to this reason. Am already using 0.3*cost for the pgRouting process

Comment (radumas) [reply gis/bylaw_text_to_centreline/README.md:264]: This section should be updated with where that issue ended: only 15 of previously 84 bylaws fail there because the cost was modified to prefer the highway matching the streetname

File: gis/bylaw_text_to_centreline/README.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+
+The function outputs a confidence level (`con`), which is defined after the `centreline_segments` variable. This value represents how close the names of the input values are to the intersections they were matched to. Specifically, this value is the sum of the levenshtien distance between the `highway2` value and the street name it matched to and the `btwn1`/`btwn2` value and the street name it matched to. It is the sum of this value for both intersections.
+
+# Quality Control
+
+There is a [Github Issue](https://github.com/CityofToronto/bdit_data-sources/issues/188) on weird bylaws that haven't been matched from Chelsea's work. \
+For example: \
+i) The streets stated in bylaws are slightly off i.e. they do not intersect or are rather far apart. \
+ii) Directionality matters. Northbound and Southbound route between two intersection are of different centrelines such as the Allen Rd. \
+iii) The error message `line_locate_point: 1st arg isn't a line` returned for 49 distinct id's. This mostly occurs when trying to slice a multilinestring.
```
Comment (radumas) [inline gis/bylaw_text_to_centreline/README.md:260]: are these points referring to the headings below?

File: gis/bylaw_text_to_centreline/README.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+## pgRouting returns the shortest path but street name different from `highway`
+
+This can be found at [issue #268](https://github.com/CityofToronto/bdit_data-sources/issues/268#issuecomment-595973836). 
+This happened for `bylaw_id` = 6577 where `highway` = 'Garthdale Court' and `between` = 'Overbrook Place and Purdon Drive'. The two intersection ids found are 13448816 and 13448300 (marked as red cross below). The blue line is the result from pg_routing whereas the highlighted yellow path indicates the road segment from the bylaw.
+
+![image](https://user-images.githubusercontent.com/54872846/76124099-f7475800-5fc7-11ea-865e-963edefa43be.png)
+
+## Direction stated on bylaws is not taken into account
+
+Which is also somehow related to problem (ii) stated above. This can be found at [issue #276](https://github.com/CityofToronto/bdit_data-sources/issues/276). For example: `highway` = 'Black Creek Drive (southbound)' and `between` = 'Eglinton Avenue West and a point 200 metres north of Weston Road'
```
Comment (radumas) [inline gis/bylaw_text_to_centreline/README.md:273]: I would just say these are the same issue and link them explicitly. The direction of the highway is mentioned in the bylaw but we don't use that and that's a problem. 

**Is there any info on how to identify if this is an issue for a set of processed bylaws?**

File: gis/bylaw_text_to_centreline/README.md [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+
+The image below shows the four possible cases (A1, A2, B1 and B2)
+![image](https://user-images.githubusercontent.com/54872846/77707649-284df380-6f9c-11ea-9509-b0db0cdf24c5.png)
+
+
+## Confidence output
+
+The function outputs a confidence level (`con`), which is defined after the `centreline_segments` variable. This value represents how close the names of the input values are to the intersections they were matched to. Specifically, this value is the sum of the levenshtien distance between the `highway2` value and the street name it matched to and the `btwn1`/`btwn2` value and the street name it matched to. It is the sum of this value for both intersections.
+
+# Quality Control
```
Comment (radumas) [inline gis/bylaw_text_to_centreline/README.md:254]: This section should list out how to identify these problems in the output. Is it because of `WARNING` messages thrown by the function? Or are some of these identified from manual inspection, but now we now why (e.g. "street name is duplicated across former municipalities")

It should also be split into issues identified and a new "Outstanding work" section referencing the open tasks that could solve some quality control problems.

**Review by radumas — CHANGES_REQUESTED** [review]
Extensive and improving. 

Note the uses of "I", nobody who reads this in the future will know who that was (and does it matter? We're not going to reach out to you with questions about those statements). Replace with either the team "we" or if it's important that someone knows who was involved in that sentence refer to yourself in the third person "Joven" (unless Chelsea wrote those sentences). 

Allllso, as part of our improving the structure of the data-sources repo. Can you include a `gis` section in the top-level/root README with a brief description of, for example

>`gis/text_to_centreline/` contains sql used to transform text description of street (in particular bylaws) into centreline geometries. See the README for details on how to use

And ditto create a README in `gis` with the same

File: gis/text_to_centreline/README.md [hunk 17]
```diff
@@ -0,0 +1,800 @@
...
+
+## Confidence output
+
+The function outputs a confidence level (`con`), which is defined after the `centreline_segments` variable. This value represents how close the names of the input values are to the intersections they were matched to. Specifically, this value is the sum of the levenshtien distance between the `highway2` value and the street name it matched to and the `btwn1`/`btwn2` value and the street name it matched to. It is the sum of this value for both intersections.
+
+# Creating Bylaws Speed Limit Layer
+
+The function created above was to read bylaws and return the centrelines involved, be it partial or complete. The following steps then have to be done in order to prepare the final bylaws speed limit layer. 
+
+## Using the Function
```
Comment (radumas) [inline gis/text_to_centreline/README.md:452]: >Run `text_to_centreline()` to generate geometries

File: gis/text_to_centreline/README.md [hunk 18]
```diff
@@ -0,0 +1,800 @@
...
+
+However, the function does not include the `date_added` and `date_repealed` which are essential in the next part. Therefore, in order to get that information, the following query has to be run too. This can be one of the [future enhancement](#Include-date_added-and-date_repealed).
+```sql
+CREATE TABLE gis.bylaws_routing_dates AS
+SELECT result.*, dates.date_added, dates.date_repealed FROM gis.bylaws_routing result
+LEFT JOIN jchew.bylaws_added_repealed_dates dates
+USING (id)
+```
+
+## Match to Centrelines and Categorize Bylaws
```
Comment (radumas) [inline gis/text_to_centreline/README.md:478]: >Incorporate Centrelines without Bylaws and Cut Centrelines

File: gis/text_to_centreline/README.md [hunk 19]
```diff
@@ -0,0 +1,800 @@
...
+
+The previous step only converts all bylaws into centrelines and do not include centrelines that are not stated in the bylaws. This [mat view query](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_centreline_categorized.sql) categorizes bylaws into different parts and incorporates that into the centreline layer into a mat view named `gis.bylaws_centreline_categorized`. We check if the centrelines are involved in any bylaws, if they are not, set the speed limit to 50km/h. If they are just partially included in the bylaws, we check if there's another bylaw that governs that centreline. If there is, apply the next bylaw; If there is none, set the speed limit to 50km/h. For a centreline that is included partially in more than one bylaws, it falls into the part two category. This query may seem long but it is technically just handling the bylaws in a few parts.
```
Comment (radumas) [inline gis/text_to_centreline/README.md]: The function categorizes centrelines based on how the processed bylaw geometries intersect with the original geometry (not all, entirely, partially, or multiple bylaws). Man this is challenging to communicate. I think a graphic could help of the different cases this is dealing with (if you have time 😨 )

File: gis/text_to_centreline/README.md [hunk 20]
```diff
@@ -0,0 +1,800 @@
...
+LEFT JOIN jchew.bylaws_added_repealed_dates dates
+USING (id)
+```
+
+## Match to Centrelines and Categorize Bylaws
+
+The previous step only converts all bylaws into centrelines and do not include centrelines that are not stated in the bylaws. This [mat view query](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_centreline_categorized.sql) categorizes bylaws into different parts and incorporates that into the centreline layer into a mat view named `gis.bylaws_centreline_categorized`. We check if the centrelines are involved in any bylaws, if they are not, set the speed limit to 50km/h. If they are just partially included in the bylaws, we check if there's another bylaw that governs that centreline. If there is, apply the next bylaw; If there is none, set the speed limit to 50km/h. For a centreline that is included partially in more than one bylaws, it falls into the part two category. This query may seem long but it is technically just handling the bylaws in a few parts. 
+
+1. no_bylaw -> centrelines not involved in bylaws and so the speed limits are set to 50 km/h
```
Comment (radumas) [inline gis/text_to_centreline/README.md:482]: >No bylaw matched, therefore default of 50 km/hr for urban street applies.

File: gis/text_to_centreline/README.md [hunk 21]
```diff
@@ -0,0 +1,800 @@
...
+
+2. whole_added -> centrelines involved in bylaws, be it fully or partially
```
Comment (radumas) [inline gis/text_to_centreline/README.md:484]: Are these numbers referring to steps in the query?

File: gis/text_to_centreline/README.md [hunk 22]
```diff
@@ -0,0 +1,800 @@
...
+
+ii) [L65](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_centreline_categorized.sql#L65): ` (centreline.fcode_desc::text = ANY (ARRAY['Collector'::character varying, ...` is used here to only include relevant centrelines from `gis.centreline`.
+
+iii) [L84](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_centreline_categorized.sql#L84) `WHERE whole_added.section IS NOT NULL AND whole_added.section <> '[0,1]'::numrange` is used to find centrelines where bylaws are only applied to a part of it.
+
+iv) [L123](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_centreline_categorized.sql#L123) `WHERE bylaws.geo_id = one.geo_id AND (bylaws.date_added < one.date_added OR bylaws.id < one.bylaw_id)` is used to find the previous bylaws according to the date_added but since not all bylaws have date_added, the id is used instead with bigger id representing more latest bylaw_id.
```
Comment (radumas) [inline gis/text_to_centreline/README.md]: >is used to find the previous bylaws according to the `date_added`. Since not all bylaws have `date_added`, the `id` is used instead with larger values representing more recently applied bylaws.

File: gis/text_to_centreline/README.md [hunk 23]
```diff
@@ -0,0 +1,800 @@
...
+
+v) [L135](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_centreline_categorized.sql#L135) `st_difference(next_bylaw.geom, st_buffer(part_one.geom, 0.00001::double precision)) AS geom,` is used to find the difference between the current and previous bylaw centrelines for part two cases. Note that st_difference does not work without the st_buffer here.
```
Comment (radumas) [inline gis/text_to_centreline/README.md]: Remove double negative
>Note that st_difference needs `st_buffer` here to work properly because...

File: gis/text_to_centreline/README.md [hunk 24]
```diff
@@ -0,0 +1,800 @@
...
+
+## Final Clean Up
+
+### *`gis.bylaws_speed_limit_layer`*
+The final bylaws speed limit layer is a table named `gis.bylaws_speed_limit_layer`. To be honest, the results produced in this step is very similar to the one from the previous step. BUT, the geom in this table is way more accurate as we are cutting the centreline based on the information from section whereas in the previous process, we used buffer to do the slicing. Therefore, even though there are only 303 different rows (only the geom is slightly different) comparing this mat view and the mat view from previous step, we will still use this mat view to ensure that the geom is exactly the same as the geom from `gis.centreline`. The query can be found [here](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/mat-view-bylaws_speed_limit_layer.sql) where the part mentioned below is the important part.
```
Comment (radumas) [inline gis/text_to_centreline/README.md]: wait why the different processes?

File: gis/text_to_centreline/README.md [hunk 25]
```diff
@@ -0,0 +1,800 @@
...
+|3762|	Traymore Cres	|1146129|	40	|13467350	|NULL|	Very High (100% match)|	highway2:...|...|	\[0,0.372174765194242)	|...|	...|	NULL	|NULL	|NULL|	NULL|
+|3763	|Traymore Cres|	1146129	|40|	13467350|	13467108|	Very High (100% match)|	highway2:...| ...|	[0.372174765194242,1]	|...|	...|	...|	NULL|	NULL	|NULL|
+|6583|	Glenvale Blvd|	127|	30|	13455526|	13455120	|Very High (100% match)|	highway2:...| ...|	NULL|	...|NULL|		...|	NULL|	01/15/2019	|NULL|	
+|NULL|Broadway Ave|	129|	50	|	NULL|NULL|NULL|NULL|...|NULL|||||NULL|NULL|								
+
+Look at Harvie Ave (which is considered at part two), the latest bylaw is applied to the centreline partially and the other part of the centreline is either filled with the previous bylaws or the speed limit is just set to 50 if there isn't any previous bylaws applied to that part of the centreline. Traymore Cres is considered as part one where the centreline is partially governed by a bylaw and the other part is governed by an older bylaw. Glencale Blvd is considered as whole_added where the whole centreline is related to a bylaw whereas Broadway Ave is considered as no_bylaw as there is no bylaw governing that centreline.
```
Comment (radumas) [inline gis/text_to_centreline/README.md:527]: Can you give this a more meaningful name than "part two"?

Split into a bullet for each different street.

File: gis/text_to_centreline/README.md [hunk 26]
```diff
@@ -0,0 +1,800 @@
...
+
+We might be able to solve it by separating that direction during the cleaning the data process and taking that into account when finding the right centrelines.
+
+## Levenshtein distance can fail for streets that have E / W
+
+This can be found at [issue #279](https://github.com/CityofToronto/bdit_data-sources/issues/279). For example, bylaw_id = 6667, highway = 'Welland Ave' and between = 'Lawrence Ave W and Douglas Avenue'. Since changing 'Lawrence Ave W' to 'Lawrence Ave E' only takes 1 step whereas changing 'Welland Ave' to 'Welland Rd' takes 3 steps, the intersection point with the less levenshtein distance is chosen instead, resulting in the segments of road that does not reflect the bylaws text.
+
+I also found that when there are three street names in the intersec5 column and two of the street names are almost the same with the slight change of E to W or vice versa (which happen particularly frequently for Yonge St like Adelaide St E / Yonge St / Adelaide St W, King St E / Yonge St / King St W, Bloor St E / Yonge St / Bloor St W etc), the intersection points found is weird because of the reason above.
+
+## Include former municipality element of the "highway" field
```
Comment (radumas) [inline gis/text_to_centreline/README.md:785]: Rename to "Duplicate street name and former municipality element ignored"

File: gis/text_to_centreline/README.md [hunk 27]
```diff
@@ -0,0 +1,800 @@
...
+```
+
+Table `jchew.bylaws_found_routes` contains 26646 rows (as a bylaw can return multiple centrelines) and looks like this. 
+|highway|between|note|bylaw_id|int_start|int_end|line_geom|seq|geo_id|lf_name|objectid|fcode|fode_desc|
+|--|--|--|--|--|--|--|--|--|--|--|--|--|
+|Warden Avenue|A point 305 metres north of Mack Avenue and St. Clair Avenue East|highway2: Warden Ave btwn1:  Mack Ave btwn2: St. Clair Ave E metres_btwn1: 305 metres_btwn2:  direction_btwn1: north direction_btwn2:|6872|13457981|13455952|...|1|112744|Warden Ave|21501|201300|Minor Arterial|
+
+Below shows a table on exactly how many bylaws failed at different stage. Please refer to the flow chart [here](#how-the-function-works) for a better picture. The csv dump [here](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/csv/failed_bylaws.csv) (also in table `jchew.failed_bylaws`) contains all bylaws that failed to be converted into centrelines and also the reason behind. There are, in total, 211 bylaws that do not get processed (note that these are all bylaws where deleted = false).
+
+|failed_reason|function|stage |# of bylaws that failed here|
```
Comment (radumas) [inline gis/text_to_centreline/README.md:666]: This should be higher up in this sub-section

File: gis/text_to_centreline/posted_speed_limit_update/README.md [hunk 28]
```diff
@@ -0,0 +1,33 @@
...
+
+
+## Table of Contents
+1. [Description of Project Problem](#description-of-project-problem)
+2. [Methodology](#Methodology)
+3. [Information on Subfolders](#Information-on-Subfolders)
+
+## Description of Project Problem
+
+We were given a file of approximately 8000 descriptions of locations on Toronto streets,
```
Comment (radumas) [inline gis/text_to_centreline/posted_speed_limit_update/README.md:11]: Why does the other README only say 5000? Because of bylaws that were repealed?

File: gis/text_to_centreline/README.md [hunk 29]
```diff
@@ -0,0 +1,800 @@
...
+  - [Include former municipality element of the "highway" field](#include-former-municipality-element-of-the-highway-field)
+  - [Tackle Cases with "an intersection and two offsets"](#tackle-cases-with-an-intersection-and-two-offsets)
+  - [Modify `con` (confidence level) definition to better reflect actual situation](#modify-con-confidence-level-definition-to-better-reflect-actual-situation)
+
+## Intro
+
+This is a `README` for the [`gis.text_to_centreline(bylaw_id, highway, fr, t)`](https://github.com/CityofToronto/bdit_data-sources/blob/text_to_centreline/gis/text_to_centreline/sql/function-text_to_centreline.sql) function, which is written in `postgresql`. The general purpose of this function is to take an input text description of a street location in the City of Toronto, and return an output of centreline segments that match this description. The input descriptions typically state the street that the bylaw is in effect on and the two intersections between which the bylaw. For example, you could use the function to get the centreline segments of Bloor Street between Royal York Road and St George Street.
+
```
Comment (radumas) [inline gis/text_to_centreline/README.md:41]: This section should include a brief explanation that the `posted_speed_limit_update` includes past work on transforming bylaws into centreline segments, including python code to transform the xml bylaws on Open Data into Postgresql to process.

**Review by radumas — CHANGES_REQUESTED** [review]
Looking better. In general in a README, include tables of results earlier in the text.

File: gis/text_to_centreline/README.md [hunk 30]
```diff
@@ -0,0 +1,800 @@
...
+```sql
+SELECT * FROM gis.text_to_centreline
+(6668, 'Brimorton Drive', 'A point 101 metres west of Amberjack Boulevard and a point 103 metres east of Dolly Varden Boulevard', NULL)
+```
+and the result looks like this
+
+![](jpg/case2.JPG)
+
+#### The logic behind dealing with geom to be trimmed
+To explain some logic used in the function `gis._centreline_case2()`. The part where applicable is shown below. I am using `pgrout_centreline` as the baseline and then adding or trimming centrelines where applicable according to the bylaws. If addition of centrelines is needed, then `ST_Union` is used (see A1 & A2); if subtraction/trimming of centrelines is needed, then `ST_Difference` is used (see B1 & B2).
```
[the remark this answers is not in the document]
Comment (radumas) [reply gis/text_to_centreline/README.md]: So yes the numbers A1, A2 are related to the direction the line was drawn.

**Comment by jovenc** [comment]
Above PR comments are addressed. It is ready to be reviewed 2.0 .

File: gis/text_to_centreline/sql/mat-view-bylaws_speed_limit_layer.sql [hunk 31]
```diff
@@ -278,9 +293,10 @@ UNION
     part_two_without_bylaw.date_added,
     part_two_without_bylaw.date_repealed
    FROM part_two_without_bylaw
+   JOIN gis.centreline cl USING (geo_id, lf_name)
 WITH DATA;
 
-COMMENT ON MATERIALIZED VIEW gis.bylaws_centreline_categorized
+COMMENT ON MATERIALIZED VIEW gis.bylaws_speed_limit_layer
```
Comment (radumas) [inline gis/text_to_centreline/sql/mat-view-bylaws_speed_limit_layer.sql:299]: The start of this comment should confirm that this is the layer one is expecting: Speed limits based on bylaws.

**Review by radumas — APPROVED** [review]
